File: Source/Fantasy.Jobs/Resources/ResourceManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Diagnostics;
using System.Threading;
using Fantasy.Jobs.Management;
using Fantasy.ServiceModel;

namespace Fantasy.Jobs.Resources
{
    public class ResourceManager : AbstractService, IResourceManager
    {
        protected IResourceProvider[] _providers;

        private List<ResourceGroup> _allocatedResources = new List<ResourceGroup>();

        private  Thread _checkThread;

        private void CheckJobsExist()
        {
            IJobController controller = this.Site.GetRequiredService<IJobController>();
            TimeSpan timeout = new TimeSpan(0, 5, 0);
            ILogger logger = this.Site.GetService<ILogger>();
            while (true)
            {
                Thread.Sleep(15000);
                DateTime now = DateTime.Now;
                foreach (ResourceGroup resource in _allocatedResources.ToArray())
                {
                    try
                    {
                        if (!controller.IsJobProccessExisted(resource.JobId) )
                        {
                            if (logger != null)
                            {
                                logger.LogMessage("ResourceManager", "Job {0} has exited, force release resource.", resource.JobId); 
                            }

                            Release(resource);
                        }
                    }
                    catch
                    {

                    }
                }
            }

        }

        public override void InitializeService()
        {
            _providers = AddIn.CreateObjects<IResourceProvider>("jobService/resources/provider");
            foreach (IResourceProvider provider in this._providers)
            {
                if (provider is IObjectWithSite)
                {
                    ((IObjectWithSite)provider).Site = this.Site; 
                }
                provider.Available += new EventHandler(ProviderResourceAvailable);
                provider.Revoke += new EventHandler<ProviderRevokeArgs>(ProviderResourceRevoke);
                provider.Initialize();
            }

            base.InitializeService();

            _checkThread = ThreadFactory.CreateThread(this.CheckJobsExist).WithStart();
           
           
        }


        private int _availableLock = 0;
        private object _availableSyncRoot = new object();

        private void LockAvaialbe()
        {
            lock (_availableSyncRoot)
            {
                _availableLock++;
            }
        }

        private void UnlockAvaiable()
        {
            bool fire = false;
            lock (_availableSyncRoot)
            {
                if (_availableLock > 0)
                {
                    _availableLock--;
                    fire = _availableLock == 0;
                };

            }
            if (fire)
            {
                this.OnAvailable(EventArgs.Empty);
            }
        }

        void ProviderResourceRevoke(object sender, ProviderRevokeArgs e)
        {
            this.LockAvaialbe();
            try
            {
                IGrouping<Guid, ResourceGroup>[] groups;
                lock (_resSyncRoot)
                {
                    groups = (from r in this._allocatedResources where r.ContainsResource(e.Resource) group r by r.JobId into g select g).ToArray();

                }

                Parallel.ForEach(groups, g =>
                //foreach(IGrouping<Guid, ResourceGroup> g in groups)
                {
                    foreach (ResourceGroup r in g)
                    {
                        IResourceManagerHandler handler = null;
                        lock (_handlers)
                        {
                            handler = this._handlers.GetValueOrDefault(g.Key);
                        }
                        if (handler != null)
                        {
                            if (!r.Released)
                            {
                                handler.Revoke(r.Id);
                            }
                        }
                        else
                        {
                            break;
                        }
                    }

                });
            }
            finally
            {
                this.UnlockAvaiable(); 
            }
            
        }

        void ProviderResourceAvailable(object sender, EventArgs e)
        {
            this.LockAvaialbe();
            this.UnlockAvaiable();
            
        }


        public event EventHandler Available;

        protected virtual void OnAvailable(EventArgs e)
        {
            if (this.Available != null)
            {
                this.Available(this, e);
            }
        }


        private void Release(ResourceGroup res)
        {
            this.LockAvaialbe();
            try
            {
                
                this._allocatedResources.Remove(res);
                res.Release();
            }
            finally
            {
                this.UnlockAvaiable();
            }
        }


        public virtual void Release(Guid id)
        {
            lock (_resSyncRoot)
            {
                ResourceGroup res = this._allocatedResources.Where(r => r.Id == id).SingleOrDefault();
                if (res != null)
                {
                    Release(res);
                }
            }
        }

        protected object _resSyncRoot = new object();

        public override void UninitializeService()
        {
            _checkThread.Abort();
            base.UninitializeService();
            foreach (IResourceProvider provider in this._providers)
            {
                if (provider is IDisposable)
                {
                    ((IDisposable)provider).Dispose();
                }
            }
        }

        public virtual  Guid Request(Guid jobId, ResourceParameter[] parameters)
        {
            Guid rs = Guid.Empty;
            if (parameters == null)
            {
                throw new ArgumentNullException("parameters");  
            }
            lock (this._resSyncRoot)
            {
                ResourceGroup group = new ResourceGroup(jobId, parameters);
                group.Site = this.Site;
                try
                {
                    
                    bool available = true;
                    var query = from param in parameters
                                from provider in this._providers where provider.CanHandle(param.Name)
                                select new { param = param, provider = provider };
                    foreach (var test in query)
                    {
                        object resource = null;
                        available = test.provider.Request(test.param, out resource);
                        if (available)
                        {
                            if (resource != null)
                            {
                                group.AddResource(test.provider, resource);
                            }
                        }
                        else
                        {
                            break;
                        }
                        
                    }
                    if (available)
                    {
                        rs = group.Id;
                       
                        this._allocatedResources.Add(group);
                    }
                    else
                    {
                        group.Release();
                    }
                }
                    
                catch(Exception error)
                {
                    if( ! (error is ThreadAbortException))
                    {
                        ILogger logger = this.Site.GetService<ILogger>();
                        if (logger != null)
                        {
                            logger.LogError("ResourceManager", error, "An error occurs when request a resource.");
                        }
                    }
                    group.Release();
                }
            }

            return rs;
        }

       

        
        public bool IsAvailable(ResourceParameter[] parameters)
        {
            lock (_resSyncRoot)
            {
                var query = from param in parameters
                            from provider in this._providers where provider.CanHandle(param.Name)
                            select new { param = param, provider = provider };
                foreach (var test in query)
                {
                    if (test.provider.IsAvailable(test.param) == false)
                    {
                        return false;
                    }
                }

                return true;
            }
        }

        private Dictionary<Guid,IResourceManagerHandler> _handlers = new Dictionary<Guid,IResourceManagerHandler>();

        public void RegisterHandler(IResourceManagerHandler handler)
        {
            lock (_handlers)
            {
                _handlers.Add(handler.Id(), handler);
            }
        }

        public void UnregisterHandler(Guid id)
        {
            lock (_handlers)
            {
                _handlers.Remove(id);
            }
        }

        
    }
}

File: branches/Stheno/AutoConverted/Fantasy.Jobs/Resources/ResourceManager.java
package Fantasy.Jobs.Resources;

import Fantasy.Jobs.Management.*;
import Fantasy.ServiceModel.*;

public class ResourceManager extends AbstractService implements IResourceManager
{
	protected IResourceProvider[] _providers;

	private java.util.ArrayList<ResourceGroup> _allocatedResources = new java.util.ArrayList<ResourceGroup>();

	private Thread _checkThread;

	private void CheckJobsExist()
	{
		IJobController controller = this.Site.<IJobController>GetRequiredService();
		TimeSpan timeout = new TimeSpan(0, 5, 0);
		ILogger logger = this.Site.<ILogger>GetService();
		while (true)
		{
			Thread.sleep(15000);
			java.util.Date now = new java.util.Date();
			for (ResourceGroup resource : _allocatedResources.toArray(new ResourceGroup[]{}))
			{
				try
				{
					if (!controller.IsJobProccessExisted(resource.getJobId()))
					{
						if (logger != null)
						{
							logger.LogMessage("ResourceManager", "Job {0} has exited, force release resource.", resource.getJobId());
						}

						Release(resource);
					}
				}
				catch (java.lang.Exception e)
				{

				}
			}
		}

	}

	@Override
	public void InitializeService()
	{
		_providers = AddIn.<IResourceProvider>CreateObjects("jobService/resources/provider");
		for (IResourceProvider provider : this._providers)
		{
			if (provider instanceof IObjectWithSite)
			{
				((IObjectWithSite)provider).Site = this.Site;
			}
//C# TO JAVA CONVERTER TODO TASK: Java has no equivalent to C#-style event wireups:
			provider.Available += new EventHandler(ProviderResourceAvailable);
//C# TO JAVA CONVERTER TODO TASK: Java has no equivalent to C#-style event wireups:
			provider.Revoke += new EventHandler<ProviderRevokeArgs>(ProviderResourceRevoke);
			provider.Initialize();
		}

		super.InitializeService();

		_checkThread = ThreadFactory.CreateThread(this.CheckJobsExist).WithStart();


	}


	private int _availableLock = 0;
	private Object _availableSyncRoot = new Object();

	private void LockAvaialbe()
	{
		synchronized (_availableSyncRoot)
		{
			_availableLock++;
		}
	}

	private void UnlockAvaiable()
	{
		boolean fire = false;
		synchronized (_availableSyncRoot)
		{
			if (_availableLock > 0)
			{
				_availableLock--;
				fire = _availableLock == 0;
			}

		}
		if (fire)
		{
			this.OnAvailable(EventArgs.Empty);
		}
	}

	private void ProviderResourceRevoke(Object sender, ProviderRevokeArgs e)
	{
		this.LockAvaialbe();
		try
		{
			IGrouping<Guid, ResourceGroup>[] groups;
			synchronized (_resSyncRoot)
			{
//C# TO JAVA CONVERTER TODO TASK: There is no Java equivalent to LINQ queries:
				groups = (from r in this._allocatedResources where r.ContainsResource(e.getResource()) group r by r.JobId into g select g).toArray();

			}

			//foreach(IGrouping<Guid, ResourceGroup> g in groups)
//C# TO JAVA CONVERTER TODO TASK: Lambda expressions and anonymous methods are not converted by C# to Java Converter:
			Parallel.ForEach(groups, g =>
			{
				for (ResourceGroup r : g)
				{
					IResourceManagerHandler handler = null;
					synchronized (_handlers)
					{
						handler = this._handlers.GetValueOrDefault(g.getKey());
					}
					if (handler != null)
					{
						if (!r.getReleased())
						{
							handler.Revoke(r.getId());
						}
					}
					else
					{
						break;
					}
				}
			}
		   );
		}
		finally
		{
			this.UnlockAvaiable();
		}

	}

	private void ProviderResourceAvailable(Object sender, EventArgs e)
	{
		this.LockAvaialbe();
		this.UnlockAvaiable();

	}


//C# TO JAVA CONVERTER TODO TASK: Events are not available in Java:
//	public event EventHandler Available;

	protected void OnAvailable(EventArgs e)
	{
		if (this.Available != null)
		{
			this.Available(this, e);
		}
	}


	private void Release(ResourceGroup res)
	{
		this.LockAvaialbe();
		try
		{

			this._allocatedResources.remove(res);
			res.Release();
		}
		finally
		{
			this.UnlockAvaiable();
		}
	}


	public void Release(Guid id)
	{
		synchronized (_resSyncRoot)
		{
//C# TO JAVA CONVERTER TODO TASK: There is no Java equivalent to LINQ queries:
//C# TO JAVA CONVERTER TODO TASK: Lambda expressions and anonymous methods are not converted by C# to Java Converter:
			ResourceGroup res = this._allocatedResources.Where(r => id.equals(r.Id)).SingleOrDefault();
			if (res != null)
			{
				Release(res);
			}
		}
	}

	protected Object _resSyncRoot = new Object();

	@Override
	public void UninitializeService()
	{
		_checkThread.stop();
		super.UninitializeService();
		for (IResourceProvider provider : this._providers)
		{
			if (provider instanceof IDisposable)
			{
				((IDisposable)provider).dispose();
			}
		}
	}

	public Guid Request(Guid jobId, ResourceParameter[] parameters)
	{
		Guid rs = Guid.Empty;
		if (parameters == null)
		{
			throw new ArgumentNullException("parameters");
		}
		synchronized (this._resSyncRoot)
		{
			ResourceGroup group = new ResourceGroup(jobId, parameters);
			group.Site = this.Site;
			try
			{

				boolean available = true;
//C# TO JAVA CONVERTER TODO TASK: There is no equivalent to implicit typing in Java:
//C# TO JAVA CONVERTER TODO TASK: There is no Java equivalent to LINQ queries:
//C# TO JAVA CONVERTER TODO TASK: This type of object initializer has no direct Java equivalent:
				var query = from param in parameters from provider in this._providers where provider.CanHandle(param.getName()) select new { param = param, provider = provider };
//C# TO JAVA CONVERTER TODO TASK: There is no equivalent to implicit typing in Java:
				for (var test : query)
				{
					Object resource = null;
					RefObject<Object> tempRef_resource = new RefObject<Object>(resource);
					available = test.provider.Request(test.param, tempRef_resource);
					resource = tempRef_resource.argvalue;
					if (available)
					{
						if (resource != null)
						{
							group.AddResource(test.provider, resource);
						}
					}
					else
					{
						break;
					}

				}
				if (available)
				{
					rs = group.getId();

					this._allocatedResources.add(group);
				}
				else
				{
					group.Release();
				}
			}

			catch(RuntimeException error)
			{
				if(! (error instanceof ThreadAbortException))
				{
					ILogger logger = this.Site.<ILogger>GetService();
					if (logger != null)
					{
						logger.LogError("ResourceManager", error, "An error occurs when request a resource.");
					}
				}
				group.Release();
			}
		}

		return rs;
	}




	public final boolean IsAvailable(ResourceParameter[] parameters)
	{
		synchronized (_resSyncRoot)
		{
//C# TO JAVA CONVERTER TODO TASK: There is no equivalent to implicit typing in Java:
//C# TO JAVA CONVERTER TODO TASK: There is no Java equivalent to LINQ queries:
//C# TO JAVA CONVERTER TODO TASK: This type of object initializer has no direct Java equivalent:
			var query = from param in parameters from provider in this._providers where provider.CanHandle(param.getName()) select new { param = param, provider = provider };
//C# TO JAVA CONVERTER TODO TASK: There is no equivalent to implicit typing in Java:
			for (var test : query)
			{
				if (test.provider.IsAvailable(test.param) == false)
				{
					return false;
				}
			}

			return true;
		}
	}

	private java.util.HashMap<Guid,IResourceManagerHandler> _handlers = new java.util.HashMap<Guid,IResourceManagerHandler>();

	public final void RegisterHandler(IResourceManagerHandler handler)
	{
		synchronized (_handlers)
		{
			_handlers.put(handler.Id(), handler);
		}
	}

	public final void UnregisterHandler(Guid id)
	{
		synchronized (_handlers)
		{
			_handlers.remove(id);
		}
	}


}